### ISSUE TITLE
Changing behavior of Im2Col and Col2Im to be inverses of each other.

### ISSUE THREAD

**Issue by IsCoelacanth** [pytorch/pytorch#21817]
## Motivation
Having im2col and col2im be inverses will make writing custom convolution and capsule models much easier. I was stuck with this issue for a very long time and eventually had to abandon the project.

**Comment by fmassa** [comment]
`col2im` corresponds to the derivative of `im2col`, and as such should be summed and not averaged.
You can always divide by the number of kernel elements to get the behavior your want.

This has been discussed in https://github.com/pytorch/pytorch/issues/20806 already, and seems to be a recurrent issue.

As such, I'm tagging it as a needs discussion. I could involve either a documentation improvement, but I'm not sure we should add an extra option to `col2im` that computes the sum instead of the average.

**Comment by SsnL** [comment]
IMHO an option to switch between sum and avg is reasonable. On another related note, could we change im2col to a view op when viewing is possible (i.e., no padding, sufficient contiguity conditions)? It will be wayyy faster.

**Comment by IsCoelacanth** [comment]
I agree with with @SsnL there should be a reduction option which allows switching from the default (summing) to average for use cases that require it. 
An alternative could be to include in the documentation how to get this behavior from the current way col2im works (code example).

**Comment by ezyang** [comment]
@gchanan says, @ifedan added a thing for AvgPool which lets you specify the divisor, so you can just say, "Divide by this inside the operation". This is a little different because the default it is different.

This seems reasonable, we will accept something along those lines; either the mode, or a divisor argument.

**Comment by pearu** [comment]
im2col and col2im can be made each other inverses (or pseudoinverses when `dilation>2`) using a certain scale tensor that can be computed as
```
scales = fold(unfold(ones))
```
Then for any `input` tensor,
```
fold(unfold(input)) == scales * input
```
holds. Note that for cases with `dilation>2` the `scales` tensor may contain zeros, hence not using a divisior tensor for generality.

Example:
```
>>> output_size=(7, 8)
>>> kernel_size=(2, 3)
>>> dilation = 1
>>> 
>>> unfold = nn.Unfold(kernel_size=kernel_size, dilation=dilation)
>>> fold = nn.Fold(output_size=output_size, kernel_size=kernel_size, dilation=dilation)
>>> 
>>> input = torch.randn(1, 1, *output_size, dtype=float)
>>> norm_scales = fold(unfold(torch.ones(1, 1, *output_size, dtype=input.dtype)))
>>> norm_scales
tensor([[[[1., 2., 3., 3., 3., 3., 2., 1.],
          [2., 4., 6., 6., 6., 6., 4., 2.],
          [2., 4., 6., 6., 6., 6., 4., 2.],
          [2., 4., 6., 6., 6., 6., 4., 2.],
          [2., 4., 6., 6., 6., 6., 4., 2.],
          [2., 4., 6., 6., 6., 6., 4., 2.],
          [1., 2., 3., 3., 3., 3., 2., 1.]]]], dtype=torch.float64)
>>> abs(fold(unfold(input)) - input * norm_scales).max()
tensor(1.7764e-15, dtype=torch.float64)
```
When setting in the above `dilation=3`, for instance, we have the pseudoinverse relation:
```
>>> norm_scales
tensor([[[[1., 1., 0., 1., 1., 0., 1., 1.],
          [1., 1., 0., 1., 1., 0., 1., 1.],
          [1., 1., 0., 1., 1., 0., 1., 1.],
          [2., 2., 0., 2., 2., 0., 2., 2.],
          [1., 1., 0., 1., 1., 0., 1., 1.],
          [1., 1., 0., 1., 1., 0., 1., 1.],
          [1., 1., 0., 1., 1., 0., 1., 1.]]]], dtype=torch.float64)
>>> abs(fold(unfold(input)) - input * norm_scales).max()
tensor(0., dtype=torch.float64)
```
Conclusions:
1. For a true (pseodo)inverse relation, introducing a scalar divisior argument (similar to AvgPool as suggested by @ezyang ) would not be sufficient, in general, as the divisior would need to be different for different tensor elements. The divisor argument would need to be a tensor in most cases.
2. `fold` could have a mode option that would be used for normalizing its output (using `1/norm_scales` as divisior tensor), it would need to raise an error in the degenerate cases when `norm_scales` contains zeros.

Possible solutions:
1. Introduce `divisior` tensor argument or `mode` flag argument to `nn.Fold` that normalizes `fold` output for non-degenerate cases, otherwise, raise ZeroDivision error.
2. Document the `fold` and `unfold` pseudoinverse relation as above and let users take advantage of the relationship as they see it appropriate.
What do you think?

Btw, currently `fold/unfold` support only 4D tensor arguments. Would there be interest in generalizing the support for 5D tensors because vol2col/col2vol are available?

**Comment by ezyang** [comment]
@fmassa and @SsnL, I'd quite like to know your opinions here!

**Comment by fmassa** [comment]
I agree with the reasoning from @pearu 

To have (pseudo) inverse relation, we need a tensor divisor. This divisor is always the same for a fixed conv layout and input size.

I think that, for now, we could document the relation between fold and unfold in their respective documentations.
We can always add an option to `fold` afterwards and add this functionality if this feature becomes of broader importance. But I'd like to see some evidence first that this is needed for implementing capsules / other important work.
